libs/blockscout-service-launcher/src/tracing/request_id_layer.rs

```diff
@@ -77,3 +87,190 @@ impl<S: Subscriber + for<'lookup> LookupSpan<'lookup>> Layer<S> for RequestIdSto
+    #[derive(Clone, Default)]
+    pub(crate) struct MockMakeWriter {
+        buf: Arc<Mutex<Vec<u8>>>,
+    }
+
+    impl MockMakeWriter {
+        #[allow(unused)]
+        pub(crate) fn new(buf: Arc<Mutex<Vec<u8>>>) -> Self {
+            Self { buf }
+        }
+
+        #[allow(unused)]
+        pub(crate) fn buf(&self) -> MutexGuard<'_, Vec<u8>> {
+            self.buf.lock().unwrap()
+        }
+
+        pub(crate) fn get_string(&self) -> String {
+            let mut buf = self.buf.lock().expect("lock shouldn't be poisoned");
+            let string = std::str::from_utf8(&buf[..])
+                .expect("formatter should not have produced invalid utf-8")
+                .to_owned();
+            buf.clear();
+            string
+        }
+    }
+
+    impl<'a> tracing_subscriber::fmt::MakeWriter<'a> for MockMakeWriter {
+        type Writer = MockWriter;
+
+        fn make_writer(&'a self) -> Self::Writer {
+            MockWriter::new(self.buf.clone())
+        }
+    }
+
+    // https://github.com/tokio-rs/tracing/blob/527b4f66a604e7a6baa6aa7536428e3a303ba3c8/tracing-subscriber/src/fmt/fmt_subscriber.rs#L1304
+    fn sanitize_timings(s: String) -> String {
+        let re = Regex::new("time\\.(idle|busy)=([0-9.]+)[mµn]s").unwrap();
+        re.replace_all(s.as_str(), "timing").to_string()
+    }
+
+    fn parse_json(s: &str) -> serde_json::Value {
+        serde_json::from_str::<serde_json::Value>(s)
+            .unwrap_or_else(|_| panic!("failed to parse '{}'", s))
+    }
+
+    fn parse_captured_logs(logs: String) -> Vec<serde_json::Value> {
+        logs.split('\n')
+            .filter(|l| !l.is_empty())
+            .map(parse_json)
+            .collect()
+    }
+
+    #[test]
+    fn request_id_layer_preserves_fields() {
+        let request_id_layer = super::layer().boxed();
+        let make_writer = MockMakeWriter::default();
+        let json_layer = tracing_subscriber::fmt::layer()
+            .json()
+            .flatten_event(true)
+            .with_current_span(true)
+            .with_span_list(true)
+            .with_writer(make_writer.clone())
+            .with_level(false)
+            .with_ansi(false)
+            .with_timer(MockTime)
+            .with_span_events(FmtSpan::ENTER)
+            .boxed();
+
+        let layers = vec![request_id_layer, json_layer];
+
+        let registry = tracing_subscriber::registry().with(layers);
+
+        with_default(registry, || {
+            let span0_with_request_id = tracing::info_span!(
+                "span0_with_request_id",
+                request_id = %"02f09a3f-1624-3b1d-8409-44eff7708208"
+            );
+            let _e0 = span0_with_request_id.enter();
+            let span1 = tracing::info_span!("span1", x = 42);
+            let _e = span1.enter();
+        });
+        let actual = sanitize_timings(make_writer.get_string());
+        assert_eq!(
```

Comment:
Did the problem exist when `with_span_list(false)` was activated? And shouldn't we test it when all spans are merged instead?

Reply:
add case with `with_span_list(false)`

Reply:
`with_span_list` doesn't really affect the test and `true` only provides more info to check (just for the record)